fsspec/implementations/cache_metadata.py

```diff
@@ -0,0 +1,210 @@
+        self.cached_files: list[Detail] = [{}]
+
+    def _scan_locations(
+        self, writable_only: bool = False
+    ) -> Iterator[tuple[str, str, bool]]:
+        """Yield locations (filenames) where metadata is stored, and whether
+        writable or not.
+
+        Parameters
+        ----------
+        writable: bool
+            Set to True to only yield writable locations.
+
+        Returns
+        -------
+        Yields (str, str, bool)
+        """
+        n = len(self._storage)
+        for i, storage in enumerate(self._storage):
+            writable = i == n - 1
+            if writable_only and not writable:
+                continue
+            yield os.path.join(storage, "cache"), storage, writable
+
+    def check_file(
+        self, path: str, ignore: Callable[[Detail], bool] | None
+    ) -> Literal[False] | tuple[Detail, str]:
+        """If path is in cache return its details, otherwise return ``False``.
+
+        Takes an optional callback so that the caller can choose to ignore
+        possible matches if, for example, they are too old.
+        """
+        for (fn, base, _), cache in zip(self._scan_locations(), self.cached_files):
+            if path not in cache:
+                continue
+            detail = cache[path].copy()
+            if ignore is not None and ignore(detail):
+                continue
+            fn = os.path.join(base, detail["fn"])
+            if os.path.exists(fn):
+                return detail, fn
+        return False
+
+    def clear_expired(self, expiry_time: int) -> Iterator[str]:
+        """Remove expired metadata from the cache.
+
+        Yields names of files corresponding to expired metadata. Caller is
+        responsible for deleting these files.
+        """
+        for path, detail in self.cached_files[-1].copy().items():
+            if time.time() - detail["time"] > expiry_time:
+                fn = detail.get("fn", "")
+                if not fn:
+                    raise RuntimeError(
+                        f"Cache metadata does not contain 'fn' for {path}"
+                    )
+                fn = os.path.join(self._storage[-1], fn)
+                yield fn
+                self.cached_files[-1].pop(path)
+
+        if self.cached_files[-1]:
+            cache_path = os.path.join(self._storage[-1], "cache")
+            with atomic_write(cache_path) as fc:
+                pickle.dump(self.cached_files[-1], fc)
+
+    def close_file(self, f: Any, path: str) -> None:
+        """Perform side-effect actions when closing a cached file.
+
+        The actual closing of the file is the responsibility of the caller.
+        """
+        # File must be writeble, so in self.cached_files[-1]
+        c = self.cached_files[-1][path]
+        if c["blocks"] is not True and len(c["blocks"]) * f.blocksize >= f.size:
+            c["blocks"] = True
+
+    def empty(self) -> bool:
+        """Return ``True`` if metadata of the writable storage is empty"""
```

Comment:
Why only of the writable store?

Reply:
We only need to know if the writable store is empty following a call to `CacheMetadata.clear_cache()` as if it is clear then the caller (`CachingFileSystem.clear_expired_cache()`) deletes all of the writable cache and resets the cache access time, etc.

It is clumsy. I could just make the name better. But probably better still is to return a boolean from `CacheMetadata.clear_cache()` to tell `CachingFileSystem.clear_expired_cache()` whether the writable cache is empty. Then we wouldn't have an `empty()` method hanging around waiting to be used incorrectly.

Reply:
I like the plan you describe

Reply:
Done.